sway-lib-core/src/raw_ptr.sw

```diff
@@ -0,0 +1,67 @@
+library raw_ptr;
+
+impl raw_ptr {
+    /// Returns `true` if the pointer is null.
+    pub fn is_null(self) -> bool {
+        let addr = asm(ptr: self) { ptr: u64 };
+        __eq(addr, 0)
+    }
+
+    /// Gets the address of the pointer.
+    pub fn addr(self) -> u64 {
+        asm(ptr: self) { ptr: u64 }
+    }
+
+    /// Calculates the offset from the pointer.
+    pub fn add(self, count: u64) -> raw_ptr {
+        let addr = asm(ptr: self) { ptr: u64 };
+        let addr = __add(addr, count);
+        asm(ptr: addr) { ptr: raw_ptr }
+    }
+
+    /// Calculates the offset from the pointer.
+    pub fn sub(self, count: u64) -> raw_ptr {
+        let addr = asm(ptr: self) { ptr: u64 };
+        let addr = __sub(addr, count);
+        asm(ptr: addr) { ptr: raw_ptr }
+    }
+
+    /// Reads the given type of value from the address.
+    pub fn read<T>(self) -> T {
+        if __is_reference_type::<T>() {
+            asm(ptr: self) { ptr: T }
+        } else {
+            asm(ptr: self, val) {
+                lw val ptr i0;
+                val: T
+            }
+        }
+    }
+    
+    /// Copies `size` bytes from `self` to `dst`.
+    pub fn copy_to(self, dst: raw_ptr, count: u64) {
+        asm(dst: dst, src: self, count: count) {
+            mcp dst src count;
+        };
+    }
+    
+    /// Copies `size` bytes from `src` to `self`.
+    pub fn copy_from(self, src: raw_ptr, count: u64) {
```

Comment:
While all the methods in this file are tested (directly or indirectly), this one is not tested at all. It would be nice to have some tests for it.